File: CoreCode/Munger.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.IO;

namespace LVLTool
{
    public static class Munger
    {
        /// <summary>
        /// Throws exception if the selected file is not UCFB or not mungable
        /// (currently tga & lua).
        /// </summary>
        /// <param name="prompt">The text to prompt the user with.</param>
        /// <returns>The munged file name.</returns>
        public static string GetMungedFile( string prompt )
        {
            string retVal = null;
            OpenFileDialog dlg = new OpenFileDialog();
            dlg.RestoreDirectory = true;
            dlg.Title = prompt;
            if (dlg.ShowDialog() == DialogResult.OK)
                retVal = dlg.FileName;
            dlg.Dispose();

            if (retVal != null)
            {
                if (IsMungableFile(retVal))
                    retVal = MungeFile(retVal, Platform.None);

                VerifyUcfbFile(retVal); // throws exception if it's not a UCFB file
            }
            return retVal;
        }

        public static string EnsureMungedFile(string fileName, Platform platform)
        {
            string retVal = fileName;

            if (IsMungableFile(fileName))
                retVal = MungeFile(fileName, platform);

            VerifyUcfbFile(retVal); // throws exception if it's not a UCFB file
            return retVal;
        }

        private static bool IsMungableFile(string file)
        {
            bool retVal = false;
            int dotIndex = file.LastIndexOf('.');
            string suffix = file.Substring(dotIndex).ToLower();
            List<string> mungableFiles = new List<string>() {
                ".tga",  ".lua",
                ".mcfg", ".fx",  ".prp", ".bnd", ".snd", ".mus",
                ".combo",".sanm",".hud", ".cfg", ".pth", ".sky",
                ".lgt",  ".pvs", ".tsr",
            };
            retVal = mungableFiles.IndexOf(suffix) > -1;
            return retVal;
        }

        private static string MungeFile(string file, Platform platform)
        {
            string retVal = null;
            int dotIndex = file.LastIndexOf('.');
            string suffix = file.Substring(dotIndex).ToLower();
            if (!Directory.Exists(Program.TmpDir))
                Directory.CreateDirectory(Program.TmpDir);

            switch (suffix)
            {
                case ".tga": retVal = TextureMunge(file, platform); break;
                case ".lua": retVal = ScriptMunge(file);  break;

                case ".mcfg": case ".fx":  case ".prp":   case ".bnd":
                case ".snd":  case ".mus": case ".combo": case ".sanm":
                case ".hud":  case ".cfg": case ".pth":   case ".sky":
                case ".lgt":  case ".pvs": case ".tsr":
                    retVal = ConfigMunge(file);
                    break;
            }
            return retVal;
        }

        private static string TextureMunge(string file, Platform platform)
        {
            string retVal = null;
            if (File.Exists(file))
            {
                int index = file.LastIndexOf('\\');
                string tmpFile = Program.TmpDir + file.Substring(index + 1);
                File.Copy(file, tmpFile, true);
                string prog = Path.GetFullPath( Program.ModToolsDir + "\\ToolsFL\\bin\\pc_TextureMunge.exe");
                if( !File.Exists(prog))
                    prog = Path.GetFullPath( Program.ModToolsDir + "\\ToolsFL\\bin\\TextureMunge.exe");
                if (!File.Exists(prog))
                    throw new Exception(
                        "Could not find texture munge program; ensure modtools directory is set correctly");
                if( platform == Platform.None)
                    platform = PlatformPrompt.PromptForPlatform();

                if (platform != Platform.None)
                {
                    FileInfo tmpFileInfo = new FileInfo(tmpFile);
                    string args = String.Format(
                        "-sourcedir {0} -platform {1} -inputfile {2} -outputdir {3}",
                        Program.TmpDir, platform.ToString(), tmpFileInfo.Name, Program.TmpDir);

                    string programOutput = Program.RunCommand(prog, args, true);
                    string outputFile = tmpFile.Replace(".tga", ".texture");
                    if (File.Exists(outputFile))
                    {
                        retVal = outputFile;
                    }
                    Console.WriteLine(programOutput);
                }
            }
            return retVal;
        }
        /*
        scriptmunge -sourcedir . -platform xbox -inputfile addme.lua -outputdir munged\ -checkdate -continue
        copy munged\addme.script ..\_LVL_XBOX\
         */
        private static string ScriptMunge(string file)
        {
            string retVal = null;
            if (File.Exists(file))
            {
                int index = file.LastIndexOf('\\');
                string tmpFile = Program.TmpDir + file.Substring(index + 1);
                File.Copy(file, tmpFile, true);
                string prog = Path.GetFullPath( Program.ModToolsDir + "\\ToolsFl\\bin\\ScriptMunge.exe");
                string args = String.Format(
                    // for scripts, platform doesn't matter; just use pc here
                    "-sourcedir {0} -platform pc -inputfile {1} -outputdir {2}",
                    ".",//Program.TmpDir, 
                    tmpFile, Program.TmpDir);

                string programOutput =  Program.RunCommand(prog, args, true);
                string outputFile = tmpFile.Replace(".lua", ".script");
                if (File.Exists(outputFile))
                {
                    retVal = outputFile;
                }
                Console.WriteLine(programOutput);
            }
            return retVal;
        }
        /// <summary>
        /// Munges the Given File (ConfigMunge).
        /// </summary>
        /// <remarks>Check command terminal for error messages when munging the file.</remarks>
        /// <param name="file">Path of the file to munge</param>
        /// <returns>The path to the munged file, null if there was an error munging the file.</returns>
        private static string ConfigMunge(string file)
        {
            string retVal = null;
            if (File.Exists(file))
            {
                int index = file.LastIndexOf('\\');
                string tmpFile = Program.TmpDir + file.Substring(index + 1);
                File.Copy(file, tmpFile, true);
                string prog = Path.GetFullPath(Program.ModToolsDir + "\\ToolsFl\\bin\\ConfigMunge.exe");
                string args = GetConfigMungeArgs(tmpFile);
                string outputExt = ".config";
                int extIndex = args.IndexOf("-ext ");
                if (extIndex > -1)
                {
                    // this functionality relies on the '-ext' arg being the last arg.
                    outputExt = "." + args.Substring(extIndex+5).Trim();
                }
                if(Program.Verbose)
                    Console.WriteLine("info: ConfigMunge; outputExt= " + outputExt);

                string programOutput = Program.RunCommand(prog, args, true);
                FileInfo fi = new FileInfo(file);
                string inputExt = fi.Extension;
                if (Program.Verbose)
                    Console.WriteLine("info: ConfigMunge; inputExt= " + inputExt);
                string outputFile = tmpFile.Replace(inputExt, outputExt);
                if (Program.Verbose)
                    Console.WriteLine("info: ConfigMunge; outputFile= " + outputFile);
                if (File.Exists(outputFile))
                {
                    retVal = outputFile;
                }
                Console.WriteLine(programOutput);
            }
            return retVal;
        }

        /// <summary>
        /// Gets the arguments for the given file type; based on the BF1 and BF2 usage of 'ConfigMunge.exe' in
        /// the batch files.
        /// </summary>
        public static string GetConfigMungeArgs(string fileName)
        {
            List<String> bf1_hash_strings_files = new List<string>(){ ".mcfg", ".snd", ".mus", ".prp", ".bnd", ".tsr" };
            List<String> bf2_hash_strings_files = new List<string>() { ".mcfg",".snd", ".mus", ".prp", ".bnd" };

            Dictionary<string, string> bf1_chunkids = new Dictionary<string,string> {
                            {".sky", "sky"}, {".prp", "prp"}, {".fx","fx"},
                            {".bnd", "bnd"},
            };
            Dictionary<string, string> bf2_chunkids = new Dictionary<string, string> {
                            {".cfg", "load"}, {".pth", "path"},  {".sky", "sky"},
                            {".fx","fx"},     {".prp", "prp"},   {".bnd", "bnd"},
                            {".lgt","lght"},  {".pvs", "PORT"}
            };

            Dictionary<string, string> bf1_exts = new Dictionary<string, string> {
                            {".fx", "envfx"}, {".prp", "prop"}, {".bnd", "boundary"}
            };
            Dictionary<string, string> bf2_exts = new Dictionary<string, string> {
                            {".cfg", "config"},   {".pth", "path"}, {".fx", "envfx"}, {".prp", "prop"},
                            {".bnd", "boundary"}, {".lgt", "light"},{".pvs", "povs"}
            };
            FileInfo fi = new FileInfo(fileName);
            string endsWith = fi.Extension;
            string retVal = string.Format("-inputfile *{0} -continue -platform pc -sourcedir {1} -outputdir {1} ",
                endsWith, Program.TmpDir);

            if (IsCurrentConfig_BF2())
            {
                if (bf2_hash_strings_files.IndexOf(endsWith) > -1)
                    retVal += "-hashstrings ";
                if (bf2_chunkids.ContainsKey(endsWith))
                    retVal += string.Format("-chunkid {0} ",bf2_chunkids[endsWith]);
                if (bf2_exts.ContainsKey(endsWith))
                    retVal += string.Format("-ext {0} ", bf2_exts[endsWith]); // the '-ext' arg needs to be last.
            }
            else
            {
                if (bf1_hash_strings_files.IndexOf(endsWith) > -1)
                    retVal += "-hashstrings ";
                if (bf1_chunkids.ContainsKey(endsWith))
                    retVal += string.Format("-chunkid {0} ", bf1_chunkids[endsWith]);
                if (bf1_exts.ContainsKey(endsWith))
                    retVal += string.Format("-ext {0} ", bf1_exts[endsWith]); // the '-ext' arg needs to be last.
            }
            return retVal;
        }

        private static bool IsCurrentConfig_BF2()
        {
            // only BF2 has the 'swbf2rm' program (whatever that is).
            string check_for = Program.ModToolsDir + "\\ToolsFL\\bin\\swbf2rm.exe";
            bool retVal = File.Exists(check_for);
            return retVal;
        }


        public static void VerifyUcfbFile(string filename)
        {
            FileStream fs = new FileStream(filename, FileMode.Open);
            byte[] buff = new byte[8];
            fs.Read(buff, 0, buff.Length);
            fs.Close();
            byte[] UCFB = new byte[]{(byte)'u', (byte)'c', (byte)'f', (byte)'b'};
            if (!BinUtils.BinCompare(UCFB, buff, 0))
                throw new Exception("Error! File is not UCFB! " + filename);
        }

    }
}

File: dart_translation/lib/munger.dart
/// Munging (texture, script, config) for LVLTool.
/// Translated from C# CoreCode/Munger.cs for Linux/CLI use.
/// UI (file dialogs) is not available; callers pass file paths.

import 'dart:io';
import 'dart:typed_data';

import 'bin_utils.dart';
import 'package:dart_translation/lvltool_config.dart';

final _mungableExtensions = [
  '.tga', '.lua', '.mcfg', '.fx', '.prp', '.bnd', '.snd', '.mus',
  '.combo', '.sanm', '.hud', '.cfg', '.pth', '.sky', '.lgt', '.pvs', '.tsr',
];

bool isMungableFile(String file) {
  final lastDot = file.lastIndexOf('.');
  if (lastDot < 0) return false;
  final suffix = file.substring(lastDot).toLowerCase();
  return _mungableExtensions.contains(suffix);
}

/// Returns munged file path, or same path if not mungable. Throws if not UCFB.
String ensureMungedFile(String fileName, [String platform = 'pc']) {
  String retVal = fileName;
  if (isMungableFile(fileName)) {
    retVal = mungeFile(fileName, platform)!;
  }
  verifyUcfbFile(retVal);
  return retVal;
}

String? mungeFile(String file, [String platform = 'pc']) {
  final lastDot = file.lastIndexOf('.');
  if (lastDot < 0) return null;
  final suffix = file.substring(lastDot).toLowerCase();
  if (!Directory(tmpDir).existsSync()) {
    Directory(tmpDir).createSync(recursive: true);
  }
  switch (suffix) {
    case '.tga':
      return _textureMunge(file, platform);
    case '.lua':
      return _scriptMunge(file);
    case '.mcfg':
    case '.fx':
    case '.prp':
    case '.bnd':
    case '.snd':
    case '.mus':
    case '.combo':
    case '.sanm':
    case '.hud':
    case '.cfg':
    case '.pth':
    case '.sky':
    case '.lgt':
    case '.pvs':
    case '.tsr':
      return _configMunge(file);
    default:
      return null;
  }
}

String? _textureMunge(String file, String platform) {
  if (!File(file).existsSync()) return null;
  final sep = file.contains('/') ? '/' : '\\';
  final lastSep = file.lastIndexOf(sep);
  final tmpFile = pathJoin(tmpDir, lastSep >= 0 ? file.substring(lastSep + 1) : file);
  File(file).copySync(tmpFile);
  String prog = pathJoin(modToolsDir, 'ToolsFL/bin/pc_TextureMunge.exe');
  if (!File(prog).existsSync()) {
    prog = pathJoin(modToolsDir, 'ToolsFL/bin/TextureMunge.exe');
  }
  if (!File(prog).existsSync()) {
    throw Exception(
        'Could not find texture munge program; ensure modtools directory is set correctly');
  }
  final args = '-sourcedir $tmpDir -platform $platform -inputfile ${tmpFile.split(sep).last} -outputdir $tmpDir';
  if (runCommand != null) {
    runCommand!(prog, args, true);
  }
  final outputFile = tmpFile.replaceAll('.tga', '.texture');
  return File(outputFile).existsSync() ? outputFile : null;
}

String? _scriptMunge(String file) {
  if (!File(file).existsSync()) return null;
  final sep = file.contains('/') ? '/' : '\\';
  final lastSep = file.lastIndexOf(sep);
  final tmpFile = pathJoin(tmpDir, lastSep >= 0 ? file.substring(lastSep + 1) : file);
  File(file).copySync(tmpFile);
  final prog = pathJoin(modToolsDir, 'ToolsFl/bin/ScriptMunge.exe');
  final args = '-sourcedir . -platform pc -inputfile $tmpFile -outputdir $tmpDir';
  if (runCommand != null) {
    runCommand!(prog, args, true);
  }
  final outputFile = tmpFile.replaceAll('.lua', '.script');
  return File(outputFile).existsSync() ? outputFile : null;
}

String? _configMunge(String file) {
  if (!File(file).existsSync()) return null;
  final sep = file.contains('/') ? '/' : '\\';
  final lastSep = file.lastIndexOf(sep);
  final tmpFile = pathJoin(tmpDir, lastSep >= 0 ? file.substring(lastSep + 1) : file);
  File(file).copySync(tmpFile);
  final prog = pathJoin(modToolsDir, 'ToolsFl/bin/ConfigMunge.exe');
  final args = getConfigMungeArgs(tmpFile);
  int extIndex = args.indexOf('-ext ');
  String outputExt = '.config';
  if (extIndex > -1) {
    outputExt = '.${args.substring(extIndex + 5).trim()}';
  }
  if (verbose) print('info: ConfigMunge; outputExt= $outputExt');
  if (runCommand != null) {
    runCommand!(prog, args, true);
  }
  final inputExt = file.contains('.') ? file.substring(file.lastIndexOf('.')) : '';
  final outputFile = tmpFile.replaceAll(inputExt, outputExt);
  return File(outputFile).existsSync() ? outputFile : null;
}

String getConfigMungeArgs(String fileName) {
  final bf1HashStrings = ['.mcfg', '.snd', '.mus', '.prp', '.bnd', '.tsr'];
  final bf2HashStrings = ['.mcfg', '.snd', '.mus', '.prp', '.bnd'];
  final bf1ChunkIds = {'.sky': 'sky', '.prp': 'prp', '.fx': 'fx', '.bnd': 'bnd'};
  final bf2ChunkIds = {
    '.cfg': 'load', '.pth': 'path', '.sky': 'sky', '.fx': 'fx',
    '.prp': 'prp', '.bnd': 'bnd', '.lgt': 'lght', '.pvs': 'PORT'
  };
  final bf1Exts = {'.fx': 'envfx', '.prp': 'prop', '.bnd': 'boundary'};
  final bf2Exts = {
    '.cfg': 'config', '.pth': 'path', '.fx': 'envfx', '.prp': 'prop',
    '.bnd': 'boundary', '.lgt': 'light', '.pvs': 'povs'
  };
  final ext = fileName.contains('.') ? fileName.substring(fileName.lastIndexOf('.')) : '';
  String retVal = '-inputfile *$ext -continue -platform pc -sourcedir $tmpDir -outputdir $tmpDir ';
  final isBF2 = _isCurrentConfigBF2();
  if (isBF2) {
    if (bf2HashStrings.contains(ext)) retVal += '-hashstrings ';
    if (bf2ChunkIds.containsKey(ext)) retVal += '-chunkid ${bf2ChunkIds[ext]} ';
    if (bf2Exts.containsKey(ext)) retVal += '-ext ${bf2Exts[ext]} ';
  } else {
    if (bf1HashStrings.contains(ext)) retVal += '-hashstrings ';
    if (bf1ChunkIds.containsKey(ext)) retVal += '-chunkid ${bf1ChunkIds[ext]} ';
    if (bf1Exts.containsKey(ext)) retVal += '-ext ${bf1Exts[ext]} ';
  }
  return retVal;
}

bool _isCurrentConfigBF2() {
  final checkFor = pathJoin(modToolsDir, 'ToolsFL/bin/swbf2rm.exe');
  return File(checkFor).existsSync();
}

void verifyUcfbFile(String filename) {
  final f = File(filename).openSync();
  final buff = f.readSync(8);
  f.closeSync();
  final ucfb = Uint8List.fromList([0x75, 0x63, 0x66, 0x62]);
  if (!binCompare(ucfb, buff, 0)) {
    throw Exception('Error! File is not UCFB! $filename');
  }
}
